services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/scheduler/events.py

```diff
@@ -503,114 +506,133 @@ async def action(cls, app: FastAPI, scheduler_data: SchedulerData) -> None:
         # invoke container cleanup at this point
         dynamic_sidecar_client: DynamicSidecarClient = get_dynamic_sidecar_client(app)
 
-        try:
-            await dynamic_sidecar_client.begin_service_destruction(
-                dynamic_sidecar_endpoint=scheduler_data.dynamic_sidecar.endpoint
-            )
-        except BaseClientHTTPError as e:
-            logger.warning(
-                (
-                    "Could not contact dynamic-sidecar to begin destruction of "
-                    "%s\n%s. Will continue service removal!"
-                ),
-                scheduler_data.service_name,
-                f"{e}",
-            )
-        app_settings: AppSettings = app.state.settings
-        dynamic_sidecar_settings: DynamicSidecarSettings = (
-            app_settings.DYNAMIC_SERVICES.DYNAMIC_SIDECAR
-        )
-
-        # only try to save the status if :
-        # - the dynamic-sidecar has finished booting correctly
-        # - it is requested to save the state
-        if (
-            scheduler_data.dynamic_sidecar.service_removal_state.can_save
-            and await are_all_services_present(
-                node_uuid=scheduler_data.node_uuid,
-                dynamic_sidecar_settings=app.state.settings.DYNAMIC_SERVICES.DYNAMIC_SIDECAR,
+        # A node can end up with all the services from a single study.
+        # When the study is closed, all the services will try to save their
+        # data. This causes a lot of disk and network stress.
+        # Some nodes collapse under load or behave unexpectedly.
+        lock_manager = RedisLockManager.instance(app)
+        assert scheduler_data.dynamic_sidecar_node_id  # nosec
+        lock: Optional[Lock] = await lock_manager.acquire_lock(
+            scheduler_data.dynamic_sidecar_node_id
+        )
+        if lock is None:
+            node_slots = await lock_manager.get_node_slots(
+                scheduler_data.dynamic_sidecar_node_id
             )
-        ):
-            dynamic_sidecar_client = get_dynamic_sidecar_client(app)
-            dynamic_sidecar_endpoint = scheduler_data.dynamic_sidecar.endpoint
+            logger.debug("All %s are currently being used. Skipping", f"{node_slots=}")
+            # Next observation cycle, the service will try again to save the state
+            # if it is able to acquire a lock
+            return
 
-            logger.info(
-                "Calling into dynamic-sidecar to save state and pushing data to nodeports"
-            )
+        async with auto_release(lock_manager, lock):
```

Comment:
hmm, can't we have something more straightforward?
something like
```python
try:
  async with lock_manager.lock():
    # do what you need to do under lock
except LockUnavailableError as err:
  logger.debug("All %s are currently being used. Skipping", f"{err.node_slots=}")
```

Reply:
So you would raise an error when acquiring the lock? I thought it would look odd.

With the current implementation you have the advantage of having all the locking code at the beginning of the function. But yours looks more pythonic.

I guess it feels better in the end. Will give it a go.